### ISSUE TITLE
Stack.TCPV4.write do not return not even fail asynchronously.

### ISSUE THREAD

**Issue by rixed** [mirage/mirage#622]
Sometime my calls to write just never return (neither Ok nor Error nor Lwt.fail).
Extract from the logs:

```
Tcp.Segment: TCP retransmission on timer seq = 466466133
Tcp.Segment: TCP retransmission on timer seq = 466466133
Tcp.Segment: TCP retransmission on timer seq = 466466133
Tcp.Segment: TCP retransmission on timer seq = 466466133
Tcp.Segment: TCP retransmission on timer seq = 466466133
Tcp.Segment: TCP retransmission on timer seq = 466466133
Tcp.Segment: Max retransmits reached for connection - terminating
```

I suspect in this particular instance that this is a MTU issue.
Also, I'm a bit concerned that this write, so early in the application (the program wrote ~2KiB in this socket before this occurred). So it seems the TCP output buffer is really really small. To be honest, even with no output buffer at all I would expect the write to still not block before the number of bytes in flight is above a threshold way greater than 2KiB, which makes me wonder if the write is maybe waiting for the acknowledgment of the passed data??

**Comment by samoht** [comment]
Which version of `mirage-tcpip` are you using?

**Comment by rixed** [comment]
Apparently none:

```
% opam list -i mirage*
# Installed packages for stubbic-flambda:
mirage              2.9.1  The MirageOS library operating system
mirage-bootvar-xen  0.3.2  Library for reading MirageOS unikernel boot parameters in Xen
mirage-clock-unix   1.0.0  A Mirage-compatible Clock library for Unix
mirage-clock-xen      1.1  A Mirage-compatible Clock library for Xen
mirage-console      2.1.3  A Mirage-compatible Console library for Xen and Unix
mirage-dns          2.5.0  Virtual package for the MirageOS DNS transports
mirage-entropy-xen  0.3.0  MirageOS entropy device
mirage-logs           0.2  A reporter for the Logs library that writes log messages to stderr, using a Mirage `CLOCK` to add timestamps.
mirage-net-xen      1.6.1  Ethernet network device driver for MirageOS/Xen
mirage-profile      0.7.0  Collect profiling information
mirage-types        2.8.0  Module type definitions for Mirage-compatible applications
mirage-types-lwt    2.8.0  Lwt module type definitions for Mirage-compatible applications
mirage-unix         2.6.0  MirageOS library for Unix compilation
mirage-xen          2.6.0  MirageOS library for Xen
mirage-xen-minios   0.8.0  Xen MiniOS guest operating system library
mirage-xen-ocaml    2.6.0  MirageOS headers for the OCaml runtime
mirage-xen-posix    2.6.0  MirageOS library for posix headers
```

**Comment by rixed** [comment]
Found it:

```
% opam list -i *tcpip*
# Installed packages for stubbic-flambda:
tcpip  2.8.1  Userlevel TCP/IP stack
```

**Comment by yomimono** [comment]
Thanks a lot for this error report, @rixed! I'll try to reproduce this shortly.  Can you elaborate on your suspicion that this is an MTU issue?

**Comment by rixed** [comment]
I suspect an MTU issue because the first write of some data longer than ~1500 bytes triggers the bug, and that packet is supposed to be routed through a GRE tunnel. This tunnel is implemented by openvswitch (2.3.0 debian stable) and it's unclear to me if openvswitch will do anything else bu drop the packet on the floor if it can't pass the tunnel.

I'd like to test with a smaller MTU. I'm not using DHCP, is there a "command line" parameter like static_ip and friends to set the MTU?

**Comment by yomimono** [comment]
The Mirage TCP/IP stack currently doesn't handle writing packets larger than the MTU well at all -- see https://github.com/mirage/mirage-tcpip/issues/237 for some previous discussion, and https://github.com/mirage/mirage-tcpip/pull/243 for some work on addressing this by implementing IP fragmentation.

**Comment by rixed** [comment]
Are you suggesting that I should break my writes into chunk smaller than MTU + all packet headers?
Is this what other users are doing?

**Comment by yomimono** [comment]
From `Stack.TCPV4.writev` you *should* indeed be able to write packets of arbitrary sizes -- it's absolutely not intended that the caller need to worry about this, but the case where the MTU is greater than 1500 is not well-tested and I'm very willing to believe we have erroneous behavior in that case.  I'll refrain from further conjecture for the moment as I'm still digging around.

**Comment by yomimono** [comment]
Is your application receiving TCP connections (via `listen`) or creating them (via `create_connection`?)  Or both, and if so, do you see the same behavior in both cases?

**Comment by rixed** [comment]
This program is both SYNing and SYNACKing, but the particular socket that hangs on write is a socket that is opened by the writer.
Also, note that if I indeed split the writes into chunks smaller than 1k then the bug does not occur.

**Comment by talex5** [comment]
BTW, if you want to find out what it's waiting for, try running with tracing on:

https://mirage.io/wiki/profiling

(call `MProf.Trace.should_resolve` on the thread that isn't returning to make it highlight it in the GUI)

**Comment by yomimono** [comment]
Improvements for this code path, along with tests for the scenario described, were added in https://github.com/mirage/mirage-tcpip/pull/313 and are included in `tcpip` releases 3.1.4 and later.
